### PR TITLE
Resume training

### PR DESCRIPTION
**Pull request by charleygros**
When your 3-days training is suddenly killed by a fd:bmaer;hbùmrq,gqgj,qg error... it would be really really nice to resume it! This PR makes this dream come true!

DONE:
- add a flag `--resume` --> to use it: `ivadomed -c <config_file> --resume` --> will open the more recent checkpoint and update model, optimizer etc to resume the training
- this checkpoint file is saved when the validation loss is best than the current best validation loss
- `best_model.pt` is extracted from the last checkpoint and saved

TODO:
- add a test, otherwise @lrouhier would not be happy
- add this option to `automate_training` to be able to... resume it as well!
- test it on a real case scenario... and for this... sadly... rosenberg looks like in a mood to help us... :-(

### CODE REVIEW

File: ivadomed/training.py [hunk 1]
```diff
@@ -263,10 +282,15 @@ def train(model_params, dataset_train, dataset_val, training_params, log_directo
 
             # UPDATE BEST RESULTS
             if val_loss_total_avg < best_validation_loss:
+                # Save checkpoint
+                state = {'epoch': epoch + 1, 'state_dict': model.state_dict(),
+                         'optimizer': optimizer.state_dict(), 'tensorboard_logger': writer,
```
Comment (andreanne-lemay) [inline ivadomed/training.py:287]: I am having a mistake here: `TypeError: cannot serialize '_io.BufferedWriter' object` It seems to be due to saving the writer. There is probably a work around this, I'll look into it :)

**Comment by andreanne-lemay** [comment]
I didn't find a workaround for the SummaryWriter and surprisingly I didn't find anything online regarding this issue. It is still minor, the consequence is that we can't see the first epochs on tensorboard. For instance, I stopped the training at epoch 14, so when I restart I get something like this on tensorboard:
image

I added the scheduler, validation loss total and patience count in the checkpoint and fixed the epoch number inconsistencies when resuming a model. It works well on my end (except for what I mentioned above). Also another minor thing, even when resuming a model we get this print: `Initialising model's weights from scratch.`. I find this confusing, but very minor. I am not sure how (and if) we want to deal with this.

File: ivadomed/training.py [hunk 2]
```diff
@@ -283,23 +316,32 @@ def train(model_params, dataset_train, dataset_val, training_params, log_directo
     if model_params["name"] == "FiLMedUnet" and debugging:
         save_film_params(gammas_dict, betas_dict, contrast_list, model_params["depth"], log_directory)
 
-    try:
```
Comment (andreanne-lemay) [inline ivadomed/training.py:286]: I am not sure why we need to change these lines

File: ivadomed/training.py [hunk 3]
```diff
@@ -283,23 +316,32 @@ def train(model_params, dataset_train, dataset_val, training_params, log_directo
...
-    except:
-        # Save best model in model directory
-        best_model_path = os.path.join(log_directory, model_params["folder_name"], model_params["folder_name"] + ".pt")
-        torch.save(model, best_model_path)
-        logger.warning("Failed to save the model as '.onnx', saved it as '.pt': {}".format(best_model_path))
+    # Save best model in log directory
+    if os.path.isfile(resume_path):
+        state = torch.load(resume_path)
+        model_path = os.path.join(log_directory, "best_model.pt")
+        torch.save(state["model"], model_path)
```
Comment (andreanne-lemay) [inline ivadomed/training.py:323]: state['model'] doesn't exist here (we don't have any key named model, we only have the state dict)

File: ivadomed/training.py [hunk 4]
```diff
@@ -123,7 +145,8 @@ def train(model_params, dataset_train, dataset_val, training_params, log_directo
     begin_time = time.time()
 
     # EPOCH LOOP
-    for epoch in tqdm(range(1, num_epochs + 1), desc="Training"):
+    for epoch in tqdm(range(num_epochs), desc="Training", initial=start_epoch):
```
Comment (charleygros) [inline ivadomed/training.py:148]: @andreanne-lemay : Don't we need a `+1` here?

File: ivadomed/training.py (hunk 2 above)
Comment (charleygros) [reply ivadomed/training.py:286]: If we don't have a validation dataset, then no checkpoint file, then no best model.

File: ivadomed/training.py (hunk 3 above)
Comment (charleygros) [reply ivadomed/training.py:323]:
```suggestion
        torch.save(model.load_state_dict(checkpoint['state_dict']), model_path)
```

File: ivadomed/training.py (hunk 4 above)
Comment (andreanne-lemay) [reply ivadomed/training.py:148]: Nope :), we have the right number of epochs when using only one param in range

**Review by andreanne-lemay — APPROVED** [review]

**Pull request merged**